protocol/protocol.typeHierarchy.proposed.md

```diff
@@ -0,0 +1,77 @@
+
+#### Type Hierarchy
+
+Many language support inheritance and LSP provides retrieving the hierarchy information with the following requests.
+
+
+The `textDocument/subTypes` request is sent from the client to the server to resolve resolve the subtypes of a symbol at a given text document position. Returns the direct subtypes in no particular order.
+
+_Client Capabilities_:
+
+```ts
+  TextDocumentClientCapabilities {
+    /**
+     * Capabilities specific to the `textDocument/subTypes`
+     *  and `textDocument/superTypes
+     */
+    typeHierarchy?: {
+    /**
+     * Whether implementation supports dynamic registration.
+     */
+    dynamicRegistration?: boolean;
+  };
+```
+
+_Server Capabilities_:
+```ts
+ServerCapabilities{
+  /**
+   * The server provides type hierarchy information
+   */
+  typeHierarchyProvider?: boolean
+}
+```
+
+```ts
+/**
+ * Represents hierarchical information about programming constructs like variables, classes,
+ * interfaces etc.
+ */
+export interface SymbolNode extends SymbolInformation {
```

Comment:
As of LSP [`3.10.0`](https://microsoft.github.io/language-server-protocol/specification#version_3_10_0), we can use the `DocumentSymbol` as is, right? It has the `children` property.

Reply:
We need one structural change, and we could simplify this PR:

The `DocumentSymbol` should be 
```js
export class DocumentSymbol {
	name: string;
	detail?: string;
	kind: SymbolKind;
	deprecated?: boolean;
	range: Range | Location;
	selectionRange: Range | Location;
	children?: DocumentSymbol[];
}
```

instead of the current state:
```js
export class DocumentSymbol {
	name: string;
	detail?: string;
	kind: SymbolKind;
	deprecated?: boolean;
	range: Range;
	selectionRange: Range;
	children?: DocumentSymbol[];
}
```

When the `range` and the `selectionRange` has the type `Range` then the `DocumentUri` is inferred. For example, when calling `textDocument/documentSymbol`, the `DocumentUri` can be inferred as `DocumentSymbolParams.textDocument.uri` from the param. If the `range` and the `selectionRange` are `Location` types, then the `DocumentUri` is explicitly defined (`Location.uri`).

If we accept this structural change in the `DocumentSymbol`, we can defined our two new methods as `textDocument/subTypes ` and `textDocument/superTypes` with `DocumentSymbol[]` return type.

We can even stick to the `children` name for the `textDocument/superTypes`; the supertypes will be real children of the `DocumentSymbol` node, only the way of the traversal differs.

CC: @tsmaeder (since you have [taken over the PR](https://github.com/Microsoft/vscode-languageserver-node/pull/346#issuecomment-400379761))

Reply:
+1 See related https://github.com/Microsoft/vscode-languageserver-node/pull/420#discussion_r223275639. 
Btw. the documentation for `DocumentSymbol#selectionRange` says 
```
The range that should be selected and revealed when this symbol is being picked, e.g the name of a function.
	 * Must be contained by the `range`.
```
So with this contract it is enough if the `range` can alternatively be a  `Location`.

Reply:
Actually I would vote for adding an option `uri: URI` property to DocumentSymbol, because the property name `range` highly suggests that the datatype is a `Range`, too. Also it will not break existing clients.